listing/listing.go

```diff
@@ -0,0 +1,195 @@
+	var items []T
+	for it.HasNext(ctx) {
+		item, err := it.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, item)
+	}
+	return items, nil
+}
+
+// Use struct{} for Req for iterators that don't need any request structure.
+type iteratorImpl[Req, Resp, T any] struct {
+	// nextReq is the request to be used to fetch the next page. If nil, then
+	// there is no next page to fetch.
+	nextReq *Req
+
+	// getNextPage fetches the next page of items, returning the deserialized
+	// response and error.
+	getNextPage func(context.Context, Req) (Resp, error)
+
+	// getItems selects the items being iterated over from the response.
+	getItems func(Resp) []T
+
+	// getNextReq is used to get the next request to be used in the next page.
+	// If nil, then there is no next page to fetch.
+	getNextReq func(Resp) *Req
+
+	// currentPage is the current page of items.
+	currentPage []T
+
+	// currentPageIdx is the index of the next item from currentPage to return.
+	currentPageIdx int
+
+	// lastErr is the last error returned by getNextPage.
+	lastErr error
+}
+
+var ErrNoMoreItems = errors.New("no more items")
+
+// Returns a new iterator. The iterator will fetch the next page of items
+// lazily, when needed. nextReq is the request to be used to fetch the initial
+// page. If nil, then no page will be fetched. getNextPage fetches the next
+// page of items, returning the deserialized response and error. getItems
+// selects the items being iterated over from the response. getNextReq is used
+// to get the next request to be used in the next page. If the returned value
+// is nil, then there is no next page to fetch.
+func NewIterator[Req, Resp, T any](
+	nextReq *Req,
+	getNextPage func(context.Context, Req) (Resp, error),
+	getItems func(Resp) []T,
+	getNextReq func(Resp) *Req,
+) *iteratorImpl[Req, Resp, T] {
+	return &iteratorImpl[Req, Resp, T]{
+		nextReq:     nextReq,
+		getNextPage: getNextPage,
+		getItems:    getItems,
+		getNextReq:  getNextReq,
+	}
+}
+
+func (i *iteratorImpl[Req, Resp, T]) loadNextPageIfNeeded(ctx context.Context) error {
+	if i.currentPageIdx < len(i.currentPage) {
+		return nil
+	}
+	if i.nextReq == nil {
+		i.currentPage = nil
+		i.currentPageIdx = 0
+		return nil
+	}
+	if i.lastErr != nil {
+		return i.lastErr
+	}
+
+	// Keep loading pages while we have a next request and the current page is
+	// empty.
+	i.currentPage = nil
+	for i.nextReq != nil && len(i.currentPage) == 0 {
+		resp, err := i.getNextPage(ctx, *i.nextReq)
+		i.lastErr = err
+		if err != nil {
+			return err
+		}
+		items := i.getItems(resp)
+		i.currentPage = items
+		i.currentPageIdx = 0
+		i.nextReq = i.getNextReq(resp)
+	}
+	return nil
+}
+
+func (i *iteratorImpl[Req, Resp, T]) Next(ctx context.Context) (T, error) {
+	var t T
+	err := i.loadNextPageIfNeeded(ctx)
+	if err != nil {
+		return t, err
+	}
+	if i.currentPageIdx >= len(i.currentPage) {
+		return t, ErrNoMoreItems
```

Comment:
Can be `T{}`?

Reply:
I tried that but Go complained. I think the way to construct the zero value for a type parameter is the `var t T` syntax. (I guess you don't know that it is a structure.)

Reply:
Thanks!